File: union-find/cpp/src/WeightedQuickUnion.cpp
#include "WeightedQuickUnion.hpp"

WeightedQuickUnion::WeightedQuickUnion(int N)
: QuickUnion(N)
, _sizes(new int[N])
{
	for (int i = 0; i < N; ++i)
	{
		_sizes[i] = 1;
	}
}

WeightedQuickUnion::~WeightedQuickUnion()
{
	delete [] _sizes;
}

void WeightedQuickUnion::makeUnion(int p, int q)
{
	int p_root = _getRoot(p);
	
	int q_root = _getRoot(q);
	
	if (p_root != q_root)
	{
		if (_sizes[p_root] >= _sizes[q_root])
		{
			_roots[q_root] = p_root;
			
			_sizes[p_root] += _sizes[q_root];
		}
		
		else
		{
			_roots[p_root] = q_root;
			
			_sizes[q_root] += _sizes[p_root];
		}
	}
}

File: union-find/python/WeightedQuickUnion.py
#!/usr/local/bin/python3
# -*- coding: utf-8 -*-

import UnionFind

from QuickUnion import QuickUnion

class WeightedQuickUnion(QuickUnion):

	def __init__(self, N):
		super().__init__(N)
		self.sizes = [1 for x in range(N)]

	def union(self, p, q):
		p_root = self.getRoot(p)
		q_root = self.getRoot(q)
		
		if p_root != q_root:
			if self.sizes[p_root] >= self.sizes[q_root]:
				self.roots[q_root] = p_root
				self.sizes[p_root] += self.sizes[q_root]
			else:
				self.roots[p_root] = q_root
				self.sizes[q_root] += self.sizes[p_root]

	def __repr__(self):
		return "{}\n{}".format(self.roots, self.sizes)

def main():
	UnionFind.main(WeightedQuickUnion)

if __name__ == "__main__":
	main()
